Feature/docker use generated ip

rmelotte opened the pull request.

rmelotte committed:
tools: docker: runner: use the docker container IP

Generating an IP address manually has a few problems:
- the generated IP is not always unique, so different containers may
  end up with the same IP during the same test run.
- the IP known to docker is not used, so inspecting the container with
  "docker inspect <container>" returns an IP that isn't used by
  anything.

run.sh:
- drop the functions that generates IP addresses.
- drop the "--ipaddr" option entirely.

entrypoint.sh:
- the first parameter is the base MAC, there is no IP anymore.
- Instead of using the IP from eth for the bridge only if
  "${bridge_ip} == 0.0.0.0", do it in all cases.

test_gw_repeater.sh: if under WSL, no need to pass "--ipaddr 0.0.0.0"
anymore.

This commit requires the prplmesh-runner docker image to be updated on
Dockerhub.

Fixes https://github.com/prplfoundation/prplMesh/issues/461.
Fixes https://github.com/prplfoundation/prplMesh/issues/927.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
tools: docker: image-pull.sh: use Gitlab registry instead of Dockerhub

Changes to the runner image requires the image to be updated on
Dockerhub.

Since we anyway plan to switch to Gitlab registry as part of issue
 #1054, change image-pull.sh to use it already.

With this change, we don't break branches still using Dockerhub, while
new branches using Gitlab registry also work.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

arnout commented on tests/test_gw_repeater.sh:
```diff
@@ -55,8 +55,8 @@ check_wsl() {
     # Docker fails to route traffic from the host to the containers
     # if the IP is anything but what was allocated by the daemon
     # when the container was created
```
Is this comment still valid?

rmelotte replied:
To me it looks like it's no longer relevant, I removed it.

arnout commented on tools/docker/run.sh:
```diff
@@ -61,7 +47,7 @@ main() {
             -d | --detach)      DETACH=true; shift ;;
             -f | --force)       FORCE=true; shift ;;
-            -i | --ipaddr)      IPADDR="$2"; shift; shift ;;
+            -i | --ipaddr)      IPADDR="--ip=$2"; shift; shift ;;
```
I don't believe anyone uses this option anyway, so let's just remove it.

rmelotte replied:
Ok, done.

arnout commented on tests/test_gw_repeater.sh:
```diff
@@ -111,12 +111,10 @@ main() {
             "${rootdir}"/tools/docker/run.sh -u "${UNIQUE_ID}" ${VERBOSE_OPT} ${FORCE_OPT} ${RP_EXTRA_OPT} \
-                start-agent -d -n "${repeater}" -m aa:bb:cc:"$index$index" -- "$@"
-            index=$((index+1))
+            start-agent -d "$DELAY" -n "${repeater}" -- "$@"
```
Something funky going on here... The `-d` option is for detaching, not for DELAY. And also you removed the indentation.

Same for all other changes in this commit.

rmelotte replied:
Oops, yes that's a mistake, thanks.

arnout commented on tools/docker/image-pull.sh:
```diff
@@ -59,12 +59,12 @@ main() {
-    # The tag used in dockerhub is the base image with : converted to -
-    upstream_tag="$(echo "$IMAGE" | tr : -)"
-    info "Base docker image $IMAGE -> tag $tag"
+    # The registry used in Gitlab has the base image name appended with : converted to -
```
NACK this. At least for me, I didn't intend to drop dockerhub and switch to the gitlab registry. Instead, what I meant was to:
- dockerbuild images as part of CI;
- push them to the registry in each pipeline;
- push them to dockerhub on master.

So the image-pull.sh script wouldn't be affected. Instead, CI should not use image-pull.sh but pull from the registry "manually". Or, better yet, not pull anything explicitly and rely on automatic pulling, by passing the image name explicitly to run.sh.

But that's just my opinion, maybe dropping dockerhub is simpler and good enough.

rmelotte replied:
> Instead, what I meant was to:

    dockerbuild images as part of CI;
    push them to the registry in each pipeline;
    push them to dockerhub on master.

This is almost what I had in mind, only that last step would have been: tag the image with "latest" on Gitlab if we're on the master branch instead of pushing to dockerhub.
It doesn't really make a difference to me though, the only thing we need to push to dockerhub on master is a robot account/token if we don't already have one, but it shouldn't be a big deal.

One change that I propose to keep though, is to tag the images like `prplmesh-runner-ubuntu-18.04` instead of `prplmesh-runner:ubuntu-18
04`. 
On Gitlab since we're going to have hundreds of different tags (one for each pipeline) I think it makes things easier if the different variants are not mixed together under a single registry name.

arnout replied:
As discussed orally: we agree.

arnout reviewed (approved):
Looks good to me so far. Is there any reason why this is still draft?

rmelotte committed:
tools: docker: runner: drop the hard-coded MAC addresses

Now that we can discover MAC addresses to run the tests, we no longer
need them to be hard-coded.

tools/docker/run.sh:
- drop the "--mac" option entirely.
  We could instead pass the mac address when invoking "docker run" to
  keep the possibility to choose it. However, the MAC address would be
  assigned to eth0, not to the bridge. Should this functionality be
  required by Boardfarm or by something else, we can always add it
  back differently in the future.

entrypoint.sh:
- we no longer get a base MAC as a parameter, so drop it and use
  random ones.

tests/test_gw_repeater.sh:
- drop the hard-coded base MAC.
- "index" is no unused, so remove it.

This commit requires the prplmesh-runner docker image to be updated on
Dockerhub.

Fixes https://github.com/prplfoundation/prplMesh/issues/1113.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>
